core/src/main/java/de/danoeh/antennapod/core/service/FeedUpdateWorker.java

```diff
@@ -2,31 +2,47 @@
 
 import android.content.Context;
 import android.support.annotation.NonNull;
+import android.util.Log;
+
 import androidx.work.Worker;
 import androidx.work.WorkerParameters;
-import de.danoeh.antennapod.core.ClientConfig;
-import de.danoeh.antennapod.core.preferences.UserPreferences;
-import de.danoeh.antennapod.core.util.FeedUpdateUtils;
+
 import org.awaitility.Awaitility;
 
 import java.util.concurrent.atomic.AtomicBoolean;
 
+import de.danoeh.antennapod.core.ClientConfig;
+import de.danoeh.antennapod.core.preferences.UserPreferences;
+import de.danoeh.antennapod.core.storage.DBTasks;
+import de.danoeh.antennapod.core.util.NetworkUtils;
+
 public class FeedUpdateWorker extends Worker {
 
+    private static final String TAG = "FeedUpdateWorker";
+
+    public static final String PARAM_RUN_IMMEDIATE = "runImmediate";
+
     public FeedUpdateWorker(@NonNull Context context, @NonNull WorkerParameters params) {
         super(context, params);
     }
 
     @Override
     @NonNull
     public Result doWork() {
+        final boolean isImmediate = getInputData() != null ?
+                getInputData().getBoolean(PARAM_RUN_IMMEDIATE, false) : false;
+        Log.d(TAG, "doWork() : isImmediate = " + isImmediate);
         ClientConfig.initialize(getApplicationContext());
 
         AtomicBoolean finished = new AtomicBoolean(false);
-        FeedUpdateUtils.startAutoUpdate(getApplicationContext(), () -> finished.set(true));
+        if (NetworkUtils.networkAvailable() && NetworkUtils.isFeedRefreshAllowed()) {
+            DBTasks.refreshAllFeeds(getApplicationContext(), null, () -> finished.set(true));
+        } else {
+            Log.d(TAG, "Blocking automatic update: no wifi available / no mobile updates allowed");
```

Comment:
This will create a deadlock because the AtomicBoolean is never updated. Moving the `await().until` part into the if would help.

Reply:
Good catch - thanks.